source/playbooks/CIS/lambda/cis116.py

```diff
@@ -0,0 +1,208 @@
+# HANDLER
+#------------------------------------------------------------------------------
+def lambda_handler(event, context):
+
+    LOGGER.info(event)
+    metrics = Metrics(event)
+
+    try:
+        for finding_rec in event['detail']['findings']:
+            finding = Finding(finding_rec)
+            LOGGER.info('FINDING_ID: ' + str(finding.details.get('Id')))
+            remediate(finding, metrics.get_metrics_from_finding(finding_rec))
+    except Exception as e:
+        LOGGER.error(e)
+
+    APPLOGGER.flush() # flush the buffer to CW Logs
+
+#------------------------------------------------------------------------------
+# REMEDIATION
+#------------------------------------------------------------------------------
+def remediate(finding, metrics_data):
+
+    message = {
+        'Note': '',
+        'State': 'INFO',
+        'Account': finding.account_id,
+        'Remediation': REMEDIATION,
+        'metrics_data': metrics_data
+    }
+
+    def failed():
+        """
+        Send Failed status message
+        """
+        message['State'] = 'FAILED'
+        message['Note'] = ''
+        notify(finding, message, LOGGER, cwlogs=APPLOGGER)
+
+    # Make sure it matches - custom action can be initiated for any finding.
+    # Ignore if the finding selected and the playbook do not match
+    cis_data = finding.is_cis_ruleset()
+    if not cis_data:
+        # Not an applicable finding - does not match ruleset
+        # send an error and exit
+        LOGGER.debug('CIS 1.16: incorrect custom action selection')
+        APPLOGGER.add_message('CIS 1.16: incorrect custom action selection')
+        return
+
+    if (cis_data['ruleid'] not in ['1.16']):
+        # Not an applicable finding - does not match rule
+        # send an error and exit
+        LOGGER.debug('CIS 1.16: incorrect custom action selection')
+        APPLOGGER.add_message('CIS 1.16: incorrect custom action selection')
+        return
+
+    #==========================================================================
+    message['AffectedObject'] = AFFECTED_OBJECT
+    try:
+        sess = BotoSession(finding.account_id, LAMBDA_ROLE)
+        iam = sess.client('iam')
+    except Exception as e:
+        LOGGER.error(e)
+        failed()
+        return
+
+    # Mark the finding NOTIFIED while we remediate
+    message['State'] = 'INITIAL'
+    notify(finding, message, LOGGER, cwlogs=APPLOGGER)
+
+    try:
+        details = finding.details['Resources'][0]['Details']
+        print(details)
+
+        aws_iam_user = details['AwsIamUser']
+        managed_policies = aws_iam_user['AttachedManagedPolicies'] if 'AttachedManagedPolicies' in aws_iam_user else []
+        username = aws_iam_user['UserName']
+        user_groups = aws_iam_user['GroupList'] if 'GroupList' in aws_iam_user else []
+        inline_policies = aws_iam_user['UserPolicyList'] if 'UserPolicyList' in aws_iam_user else []
+
+        group_name = f'{username}-sharr-cis116'
```

Comment:
Good idea to use the extension `-sharr-cis116`, but it kind of leaves artefacts around.

Suggestion: check if a group exists with just the username, if not, then just use username, else add a random 8 character string, similar to how CloudFormation does with resource names, double check that doesn't exist, then use that.

Reply:
Ok

Reply:
Fixed